### DISCUSSION TITLE
Include JSON content

### DISCUSSION THREAD

**Discussion by Andre601** [fralau/mkdocs-macros-plugin#194] [Q&A] [answered]
*This is related to mkdocs/mkdocs#3487*

I have this code snippet used as a hook in mkdocs to pull Strings from a JSON file and make them available for jinja2 templating through a `json` variable.

Tho, since I started using this plugin does it seem like the hook doesn't work anymore. Probs because the variables in jinja2 are overriden by it, making the json value no longer available. This is proofen by the fact that I have a `{% if json and json.prompts %}` check with a closing `{% else %}` line to include something else and the result is ending in that else statement.

Now, since I already use this plugin could I also just move the entire code over to it.

My python code from the hook:
```py
import jinja2
import posixpath
import json

from mkdocs.structure.files import File, Files
from mkdocs.config.defaults import MkDocsConfig
from mkdocs.structure.pages import Page

def on_page_markdown(markdown: str, page: Page, files: Files, config: MkDocsConfig, **kwargs):
    if page.file.src_uri != 'twitch-minigames/common/quip-battle.md':
        return
    
    path = posixpath.sep.join([config.docs_dir, "assets/extra_quips.json"])
    with open(path, encoding="utf-8") as json_file:
        json_content = json.load(json_file)
    
    return jinja2.Template(markdown).render(json = json_content)
```
The templating used on the page:
```
{% if json and json.prompts %}
  <table>
    <thead>
      <tr>
        <th><small>Auto-generated from <a href="https://github.com/chaosleaguewiki/chaosleaguewiki.github.io/blob/main/docs/assets/extra_quips.json" target="_blank" rel="nofollow">extra_quips.json</a></small></th>
      </tr>
    </thead>
    <tbody>
      {% for entry in json.prompts %}
        <tr>
          <td>{{ entry }}</td>
        </tr>
      {% endfor %}
    </tbody>
  </table>
{% else %}
  No prompts available to display.
{% endif %}
```

What would be the best aproach to move this over to the plugin to re-add this behaviour?

**Comment by fralau** [comment, marked as answer]
If my understanding is correct, it should be straightfoward:

1. Define a macro `read_json_file()` in `main.py` which is a function that returns the  `json` variable.
2. Slam the template directly in the markdown page where it is fits.

I haven't tested but it should be along those lines:

```python
def define_env(env):
    """
    This is the hook for defining variables, macros and filters

    - variables: the dictionary that contains the environment variables
    - macro: a decorator function, to declare a macro.
    - filter: a function with one of more arguments,
        used to perform a transformation
    """

   @env.macro
   def read_json_file(): 
       "Return json variable"
        path = "assets/extra_quips.json"
        with open(path, encoding="utf-8") as json_file:
            return json.load(json_file)
```

Then your markdown file:


```
{{ json = read_json_file() }}

{% if json and json.prompts %}
  <table>
    <thead>
      <tr>
        <th><small>Auto-generated from <a href="https://github.com/chaosleaguewiki/chaosleaguewiki.github.io/blob/main/docs/assets/extra_quips.json" target="_blank" rel="nofollow">extra_quips.json</a></small></th>
      </tr>
    </thead>
    <tbody>
      {% for entry in json.prompts %}
        <tr>
          <td>{{ entry }}</td>
        </tr>
      {% endfor %}
    </tbody>
  </table>
{% else %}
  No prompts available to display.
{% endif %}
```

**Reply by Andre601** [reply]
`{{ json = read_json_file() }}` is not valid.

I assume you meant `{% set json = read_json_file() %}` here.

**Reply by Andre601** [reply]
Also, want to point out that I tried `env.config.docs_dir` which according to the docs gives the MkDocs config object to use here... And it reports issues regarding a `LegacyConfig`, so I assume the macros plugin uses a legacy config of MkDocs rather than the config instance itself?

**Comment by fralau** [comment]
> `{{ json = read_json_file() }}` is not valid.
> 
> I assume you meant `{% set json = read_json_file() %}` here.

Yes, you're right.

**Comment by fralau** [comment]
> Also, want to point out that I tried `env.config.docs_dir` which according to the docs gives the MkDocs config object to use here... And it reports issues regarding a `LegacyConfig`, so I assume the macros plugin uses a legacy config of MkDocs rather than the config instance itself?

If I am not mistaken, the environment uses `doc_dir` anyway, so there should be no need to specify it.

Where is `LegacyConfig` defined?

**Reply by Andre601** [reply]
I just used `env.config.docs_dir` in `on_pre_page_macros` and received an error mentioning LegacyConfig and that it doesn't have a `docs_dir`, making me assume that this seems to be an old thing by MkDocs for backwards compat or smth...
